src/mmds/src/token.rs

```diff
@@ -0,0 +1,547 @@
+            .map_err(|_| Error::TokenGeneration)?;
+
+        Ok((payload, tag_bytes))
+    }
+
+    /// Encode token structure into a string using base64 encoding.
+    fn base64_encode_token_struct(token: &Token) -> Result<String, Error> {
+        let token_bytes: Vec<u8> = bincode::serialize(token).map_err(|_| Error::TokenGeneration)?;
+        let mut encoded_token = Vec::new();
+        // Ensure vector has enough space for base64 encoding.
+        encoded_token.resize(get_size_of_base64_encoding(token_bytes.len()), 0);
+        // Encode token structure bytes into base64.
+        let size = base64::encode_config_slice(token_bytes, base64::STANDARD, &mut encoded_token);
+        encoded_token.resize(size, 0);
+
+        // Convert base64 bytes to String.
+        // Safe to unwrap because `encoded_token` represents a base64 encoding.
+        Ok(String::from_utf8(encoded_token).unwrap())
+    }
+
+    /// Attempts to decrypt expiry value within token sequence. Returns false if expiry
+    /// cannot be decrypted. If decryption succeeds, returns true if token has not expired
+    /// (i.e. current time is greater than expiry) and false otherwise.
+    pub fn is_valid(&self, token: &str) -> bool {
+        let mut ciphertext = Vec::new();
+
+        // Decode token struct from base64.
+        let token_struct = match TokenAuthority::base64_decode_token_struct(token) {
+            Ok(token_struct) => token_struct,
+            Err(_) => return false,
+        };
+
+        // Validate constant fields inside decoded token structure.
+        if !token_struct.sanity_check() {
+            return false;
+        }
+
+        // Construct ciphertext from payload and tag within token structure.
+        ciphertext.extend_from_slice(&token_struct.payload);
+        ciphertext.extend_from_slice(&token_struct.tag);
+
+        // Decrypt ttl using AES-GCM block cipher.
+        let expiry = match self.decrypt_expiry(&ciphertext, &token_struct.iv) {
+            Ok(expiry) => expiry,
+            Err(_) => return false,
+        };
+
+        // Get current time.
+        let time_now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        let time_now_as_seconds = time_now.as_secs();
+
+        expiry > time_now_as_seconds
+    }
+
+    /// Decrypt ciphertext composed of payload and tag to obtain the expiry value.
+    fn decrypt_expiry(&self, ciphertext: &[u8], iv: &[u8]) -> Result<u64, Error> {
+        // Create Nonce object from initialization vector.
+        let nonce = Nonce::from_slice(iv);
+        // Decrypt expiry as vector of bytes from ciphertext.
+        let expiry_as_bytes = self
+            .cipher
+            .decrypt(nonce, ciphertext)
+            .map_err(|_| Error::ExpiryExtraction)?;
+        // Get expiry value in seconds from bytes.
+        let expiry: u64 =
+            bincode::deserialize(&expiry_as_bytes).map_err(|_| Error::ExpiryExtraction)?;
```

Comment:
bincode has support for limiting the memory usage used for the deserialization. This is what we use in seccompiler for example, since the input comes from untrusted sources and deserialization is known to be a target for resource-exhaust attempts.
Example: https://github.com/firecracker-microvm/firecracker/blob/9718aa6a924d151abf8b09b2226fc76c31b51c49/src/seccompiler/src/lib.rs#L90

Reply:
That is a great call-out!

I added a size check to the beginning of the token string validation, to ensure we don't bother decoding the tokens that are not the expected size. Since we expect a string of a certain length, I think it makes more sense to ensure the input in this manner. WDYT?

Reply:
That's a good idea, but this bincode memory limit is actually defending against potential cases in the implementation where the memory taken up for the deserialisation is not proportional to the length of the bytes array, so I suggest we use it to guard against this.

It also applies to the other bincode deser use-case (for the token struct).

In this specific case of expiry_as_bytes, we may actually remove the bincode usage by splitting the u64 into an array of 4 u8s based using bitwise shifts, if that makes sense

Reply:
I removed serialization/deserialization for expiry value. We now only use (de)serialize for Token struct operations. 

I also included a limit the memory usage used for the deserialization operation. PTAL.